### PR TITLE
Popup-related changes

### LINKED ISSUE #1234: "Addonify" popups
Moved from #773

Right now, addon popups are pretty much part of the extension, instead of independent from it and simply addons.
Possible improvements are:

- [x] No need for popups to specify what strings they'll use, so that msg() calls work like in userscripts and persistent scripts
- [ ] No need for addons to set the HTML lang (could be set by the extension popup)
- [x] Shared CSS classes between settings page, extension popup and addon popups
- [ ] Way for addons to detect light mode or dark mode
- [ ] Way for popups to know if they're fullscreen through `addon.self`
- [ ] `addon.*` APIs available in popups, as well as `msg` and `safeMsg`
- [ ] Communicating to userscripts and persistent scripts
- [ ] `window.Vue` and Sora automatically available

### LINKED ISSUE #1652: Be able to add custom games to cloud games list
I'm making a multiplayer game and I'd like to keep a count of how many players are online at what time
**Describe the solution you'd like**
Be able to put custom games in the cloud games list

### PR DESCRIPTION
**Pull request by apple502j**
\- which required me to refactor the whole popups thing. Also we have popup API now. ~~And that Firefox bug is absolutely fixed, for real.~~ this is a piece of cake

Resolves #1234 

UPDATE: This PR refactors popup code and adds popup API. cloud-games was refactored so that the projects can be configured in the future, but this PR **does not add configurable projects**.

### CODE REVIEW

File: addon-api/popup/Popup.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  }
+
+  /**
+   * Whether the user has enabled light mode on Scratch Addons settings.
+   * @type {boolean}
+   */
+  get isLightMode() {
+    return scratchAddons.isLightMode;
+  }
+}
```
Comment (TheColaber) [inline addon-api/popup/Popup.js:17]: Maybe make this just `fullscreen` and `lightmode`?

File: addons/cloud-games/addon.json [hunk 2]
```diff
@@ -11,8 +11,25 @@
   ],
   "popup": {
     "icon": "../../images/icons/cloud.svg",
-    "name": "Games"
+    "name": "Games",
+    "script": "popup.js",
+    "html": "popup.html"
```
Comment (TheColaber) [inline addons/cloud-games/addon.json:16]: Make script, to an array?

To be consistent, these values should be relative to the addon's folder? I'm trying to say that you should remove the popups folder and put it's files in the addons's folders.

File: addons/cloud-games/addon.json [hunk 3]
```diff
@@ -11,8 +11,25 @@
...
   },
+  "settings": [
+    {
+      "name": "Studio URL",
+      "id": "url",
+      "type": "string",
+      "default": "https://scratch.mit.edu/studios/539952/"
```
Comment (TheColaber) [inline addons/cloud-games/addon.json:23]: Just id maybe?

File: background/get-popups.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+const POPUP_PREFIX = chrome.runtime.getURL("popups");
+
+chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
+  if (!request?.requestPopupInfo) return;
+  // For some reason non-popups managed to request popup info?
```
Comment (TheColaber) [inline background/get-popups.js:5]: Why

**Review by TheColaber — CHANGES_REQUESTED** [review]
Is it possible to somehow remove the popups' translations from _locales to addon-i10n?

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (apple502j) [reply addons/cloud-games/addon.json:16]: Not gonna happen for security reasons.

**Comment by apple502j** [comment]
Also, note that addon translations, including popup translations, are in fact managed in addons-l10n

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (apple502j) [reply addons/cloud-games/addon.json:16]: I can make scripts an array. Not sure if it's that useful tho

File: addon-api/popup/Popup.js (hunk 1 above)
Comment (apple502j) [reply addon-api/popup/Popup.js:17]: We have `isLoggedIn`. We can keep this

File: addons/cloud-games/addon.json (hunk 3 above)
Comment (apple502j) [reply addons/cloud-games/addon.json:23]: Maybe. `url` is fine IMO, though.

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (TheColaber) [reply addons/cloud-games/addon.json:16]: What reasons?

File: addons/cloud-games/addon.json (hunk 3 above)
Comment (TheColaber) [reply addons/cloud-games/addon.json:23]: Only scratch urls work, so why not just shorten it to the id?

Comment (RedGuy12) [reply addons/cloud-games/addon.json:23]: it's user-facing, and the user is more likely to know what the url is

but if we describe the id to them, they could probably be able to find it. plus that'd make it so we don't have to worry about them ommitting the trailing slash, or adding /projects to the end or using non scratch URLs

i actually like that idea better

File: background/get-popups.js (hunk 4 above)
Comment (RedGuy12) [reply background/get-popups.js:5]: he obviously doesn't know

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (apple502j) [reply addons/cloud-games/addon.json:16]: Any websites could embed popups, which is not good for obvious reasons.

File: addons/cloud-games/addon.json (hunk 3 above)
Comment (apple502j) [reply addons/cloud-games/addon.json:23]: Note that while this input says "Studio URL", any string that has digits inside (including plain studio ID) works.

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (TheColaber) [reply addons/cloud-games/addon.json:16]: wut... im asking you to remove the popups folder and move it to addons folder.

File: addons/cloud-games/addon.json (hunk 3 above)
Comment (TheColaber) [reply addons/cloud-games/addon.json:23]: Maybe `STUDIO URL/ID`

Comment (apple502j) [reply addons/cloud-games/addon.json:23]: Studio URL works IMO. In most cases they'll get an error if you put random words.

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (apple502j) [reply addons/cloud-games/addon.json:16]: And that's the consequence. Any websites could embed popups if you do that (at least on Chrome with MV2), which is not good for obvious reasons.

**Comment by apple502j** [comment]
FYI:

Popup Loader and Background messaging chart

**Comment by WorldLanguages** [comment]
I'm not sure about letting users enter other studio IDs. But addonifying popups looks great

**Comment by apple502j** [comment]
Yep. Maybe add another "censorship" to cs.js? :P

**Comment by ErrorGamer2000** [comment]
I am working on the `addon.messaging` api. I only currently have `sendMessage`(only available in userscripts) and `onMessage`(persistent scripts), though I am going to work on `connect`(userscript) and `onConnect`(persistent scripts). 

Persistent Script API:

https://github.com/ErrorGamer2000/ScratchAddons/blob/88202a25288f4976d67a493d7406e99b05e37d76/addon-api/background/MessagePasser.js#L1-L64

Userscript API:

https://github.com/ErrorGamer2000/ScratchAddons/blob/88202a25288f4976d67a493d7406e99b05e37d76/addon-api/content-script/MessagePasser.js#L1-L96

cs.js changes:

https://github.com/ErrorGamer2000/ScratchAddons/blob/bd695d29e8a04501e37595e94d469e8a94a2061a/content-scripts/cs.js#L717-L747

**Comment by TheColaber** [comment]
Why is that related?

**Comment by ErrorGamer2000** [comment]
> Why is that related?

The diagram shows messaging... oh. is that for the cs?

**Comment by TheColaber** [comment]
No, it's just popup and background communication. This isn't 2 addons communicating, it's just the background giving the popup some info.

**Comment by ErrorGamer2000** [comment]
ah. ok. sorry.

**Comment by lisa-wolfgang** [comment]
Some things I noticed:
- Only the newest 40 projects in a studio are loaded -- is this intentional? If so, it should be noted somewhere.
- When not ordered by amount of players -- i.e. when the player counts are all 0 -- the projects are ordered oldest to newest, which most users wouldn't expect to be the default behavior due to projects being displayed in the opposite direction in studios.
- The text box in the settings that contains the studio link is too small to fully display the link.
- The settings would really benefit from utilizing #2875, which would allow for easier reordering than studios allow.

**Comment by apple502j** [comment]
Note that use of studio is temporary and can be subject to changes before the full release.

File: popups/cloud-games/light.css [hunk 5]
```diff
@@ -9,11 +9,10 @@ body {
...
 
-.title {
-  color: #222;
-}
-
-.title a {
+.loading,
+.title,
+.title a,
+.username-list > a {
```
Comment (WorldLanguages) [inline popups/cloud-games/light.css:15]: Usernames should be blue.
Screenshot from v1.10.0:
image

File: addons/cloud-games/addon.json (hunk 2 above)
Comment (WorldLanguages) [reply addons/cloud-games/addon.json:16]: A single script is fine. "import" can always be used if needed.

File: background/get-popups.js (hunk 4 above)
Comment (WorldLanguages) [reply background/get-popups.js:5]: @apple502j Would be nice to know what's going on here

File: popups/scratch-messaging/popup.css [hunk 6]
```diff
@@ -6,6 +6,14 @@ button {
...
   padding: 10px;
   box-shadow: 0 2px 5px rgba(0, 0, 0, 0.25);
+  overflow: hidden;
+}
+
+.contents {
+  max-height: 95vh;
+  overflow-y: auto;
+  scrollbar-width: thin;
+  scrollbar-color: grey transparent;
```
Comment (WorldLanguages) [inline popups/scratch-messaging/popup.css:16]: We need to double check the scrollbar looks the same in Chrome and Firefox

File: popups/scratch-messaging/popup.html [hunk 7]
```diff
@@ -67,324 +66,331 @@
...
-          class="message-type-title hover-reverse"
-          @click="messageTypeExtended.stMessages = !messageTypeExtended.stMessages"
-        >
-          <div class="btn-dropdown">
-            <img src="../../images/icons/expand.svg" alt="v" :class="{'reverted': messageTypeExtended.stMessages}" />
-          </div>
-          <span class="message-type-title-text">{{ uiMessages.stMessagesMsg }}</span>
-          <span class="float-right"
-            ><img class="small-icon inverted" src="../../images/icons/notice.svg" /> {{ stMessages.length }}</span
+    <div class="contents">
```
Comment (WorldLanguages) [inline popups/scratch-messaging/popup.html:69]: Is wrapping everything this div the only change on this file?

Comment (WorldLanguages) [reply popups/scratch-messaging/popup.html:69]: This changes the position of the scrollbar, but I don't see any problem with that.

File: webpages/popup-loader.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        t.dispatchEvent(new CustomEvent(request.fireEvent.name))
+      );
+      return;
+    }
+  });
+
+  const addon = new Addon({
+    id: addonId,
+  });
+  const globalObj = Object.create(null);
```
Comment (WorldLanguages) [inline webpages/popup-loader.js:82]: Popups should not have `global`.

File: popups/scratch-messaging/popup.html (hunk 7 above)
Comment (apple502j) [reply popups/scratch-messaging/popup.html:69]: Yep

File: webpages/popup-loader.js (hunk 8 above)
Comment (apple502j) [reply webpages/popup-loader.js:82]: For consistency :P

File: popups/scratch-messaging/popup.css (hunk 6 above)
Comment (apple502j) [reply popups/scratch-messaging/popup.css:16]: LGTM

**Comment by apple502j** [comment]
Update: This PR is just for popup API now. No new settings. cloud-games was refactored however.

**Pull request merged**